src/libaktualizr/primary/reportqueue.cc

```diff
@@ -0,0 +1,52 @@
+#include "reportqueue.h"
+
+#include <chrono>
+
+#include "config/config.h"
+
+ReportQueue::ReportQueue(const Config& config_in, HttpInterface& http_client)
+    : config(config_in), http(http_client), shutdown_(false) {
+  thread_ = std::thread(std::bind(&ReportQueue::run, this));
+}
+
+ReportQueue::~ReportQueue() {
+  shutdown_ = true;
+  {
+    std::unique_lock<std::mutex> thread_lock(thread_mutex_);
+    cv_.notify_all();
+  }
+  thread_.join();
+}
+
+void ReportQueue::run() {
+  while (!shutdown_) {
+    bool more = false;
+    std::unique_lock<std::mutex> thread_lock(thread_mutex_);
+    if (!report_queue_.empty()) {
+      Json::Value report_array(Json::arrayValue);
+      std::lock_guard<std::mutex> queue_lock(queue_mutex_);
+      report_array.append(*report_queue_.front());
+      HttpResponse response = http.post(config.tls.server + "/events", report_array);
```

Comment:
That is true, that is not a good scenario. I actually wanted to implement sending several download reports at once (I'm told the server should handle that), but I got into a mess with how to handle multiple the fact that I don't want to take things out of the queue until I know they've been sent. I suppose I could just iterate through the queue, copy the messages, and then delete them if they are successfully sent.

Unfortunately, I don't think the swap will work. For one thing, reports is a Json::arrayValue and report_queue is a queue<unique_ptr<Json::Value>>. For another, what do I do if the send fails? (I suppose I could just hang on to the Json::arrayValue until I send it successfully.)

Reply:
@cajun-rat but the report is only popped of the queue if the http call succeeded. What's the rest of the algorithm? Wouldn't it reorder reports?

Reply:
maybe define `Json::Value report_array(Json::arrayValue)` outside of the while loop and it will continue to grow until it successfully POSTS. If POST works, clear the array and start over? Then the locking could just be:
```
{
  std::lock_guard<std::mutex> queue_lock(queue_mutex_); 
  report_array.append(*report_queue_.front());
}
```

Reply:
Sorry, I didn't explain my suggestion very well.  I was actually proposing that the report_queue_ could become a std::vector, because we would only be appending to the end of it.  The idea was to grab the mutex and take a whole batch of messages from the queue, then release the mutex and process them.

If the send fails, then would would grab the mutex, swap the things we took back, then copy any new items to the end of the list (which will be empty in the common case).

We would have to repack the std::vector into a Json Array, as part of the sending process.

Reply:
After some contemplation and discussion, I think @doanac's solution is probably the simplest and thus the best. (Thanks for the suggestion!)

Reply:
Agreed!